Add HTML repr for groupby dataframe and series

JulianWgs opened the pull request:
After I saw the representation of a group by in Airtable, I wanted something similar in pandas. It always bothered me that there was no representation for a group by.

Update: 25.07.2020:
These are old screenshots, but still resemble the look. A new addition is the limitation of groups displayed. I don't know how to make the displaying width across all groups the same. If anyone has an idea, I'd appreciate it.

Update: 26.07.2020:
I've added three tests:

1. All rows in groups and all groups are shown
2. All groups are shown, but not all rows in groups.
3. All rows in groups are shown, but no all groups.

I unsure if it wise to put all cases into one test with an if statement which checks which of the above cases it is. Feedback welcome.

**DataFrame**:
![Screenshot from 2020-06-22 03-13-44](https://user-images.githubusercontent.com/31596773/85239940-75742100-b436-11ea-8b6a-99670d93591a.png)

**Series**:
![Screenshot from 2020-06-22 03-14-46](https://user-images.githubusercontent.com/31596773/85239970-8fadff00-b436-11ea-8f9e-b103c1a3ec2b.png)

This is a WIP implementation. Open issues are:
- [x] improve variable names
- [x] produce more clean HTML code
- [x] limit number of maximum group by DataFrames displayed (similar to the row limit)
- [x] improve look (Did not improve it, but I'm with it)
- [ ] Same width across all groups


- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry


Truncated groups with a maximum of 3 groups:
![Screenshot from 2020-09-05 16-17-33](https://user-images.githubusercontent.com/31596773/92306995-79becd80-ef93-11ea-86ca-d042ff2116b9.png)

JulianWgs committed:
Add html repr for groupby dataframe and series

JulianWgs committed:
Sort imports with isort in groupby.py

simonjayhawkins commented:
@JulianWgs Are you still working on this?

JulianWgs commented:
Is this something of interest? 

1) I'm not sure if the added complexity of the code is worth the cleaner html.
2) Also I would need some help on how to improve the visual quality (design hints not necessarly technical help). The only thing I can think of is to set the column width equally across all tables.

simonjayhawkins commented:
I would suggest that if you were expecting feedback on the code, that you mark the PR as ready for review. Alternatively, if you want feedback on the idea, that you first open a feature request issue (and fill out the template)

JulianWgs committed:
Improve variable naming

JulianWgs committed:
Add display.max_groups to config

JulianWgs committed:
Add group display truncation for too many groups

For 20,000 groups the function takes 260 ms ± 12.2 ms

- Alternative approach, how to get in the dots in the middle elegantly?

def _repr_html_(self) -> str:
    group_names = list(self.groups.keys())
    max_groups = get_option("display.max_groups")
    if max_groups < self.ngroups:
        n_start = (max_groups + 1) // 2
        n_end = max_groups - n_start
        group_names = group_names[:n_start] + group_names[-n_end:]
    repr_html = ""
    for group_name in group_names:
        group = self.groups[group_name]
        if not hasattr(group, "to_html"):
            group = group.to_frame()
        repr_html += f"<H3>Group Key: {group_name}<H3/>"
        repr_html += group.to_html(
            max_rows=get_option("display.max_rows") // self.ngroups
        )
    return repr_html

pep8speaks commented:
Hello @JulianWgs! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/config_init.py`](https://github.com/pandas-dev/pandas/blob/0068e5b71c277d84b7c9cef6fee2ee4531bf3eed/pandas/core/config_init.py):

> [Line 96:77](https://github.com/pandas-dev/pandas/blob/0068e5b71c277d84b7c9cef6fee2ee4531bf3eed/pandas/core/config_init.py#L96): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 345:89](https://github.com/pandas-dev/pandas/blob/0068e5b71c277d84b7c9cef6fee2ee4531bf3eed/pandas/core/config_init.py#L345): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 88 characters)

JulianWgs committed:
Implement faster and more scalable list variant

- takes 8.97 ms ± 352 µs for 20,000 groups and is not that dependent on
numer of groups

JulianWgs committed:
Black config_init

JulianWgs committed:
Fix bug which displayed too few rows

- max rows per group was calculated with the number of groups in the
groupby and did not consider max_groups setting

JulianWgs committed:
Add test for groupby representation

- load dataframes from html output and compare them

JulianWgs committed:
Delete trailing whitespace in comment

- https://github.com/pandas-dev/pandas/pull/34926/checks?check_run_id=909776147#step:6:16
- black did not pick it up

JulianWgs committed:
Add test cases for truncated rows and groups

JulianWgs commented:
@simonjayhawkins How do I fix the "ImportError: lxml not found, please install it" error in Travis CI ([Link](https://travis-ci.org/github/pandas-dev/pandas/jobs/711891902#L1318))?

JulianWgs commented:
@simonjayhawkins

JulianWgs commented:
@simonjayhawkins

simonjayhawkins commented:
sorry. will look soon. in the meantime merging upstream/master to resolve conflicts may help

JulianWgs committed:
Merge remote-tracking branch 'upstream/master'

JulianWgs committed:
Skip test if lxml is not installed

JulianWgs commented:
The problem was that the [lxml](https://lxml.de/) dependency was missing. I've added `@td.skip_if_no("lxml")` to skip the test if it is missing. This is also done [here](https://github.com/pandas-dev/pandas/blob/497ede8065e81065636908fecd31432a91c1ff64/pandas/tests/io/test_html.py#L79).

Please review the PR :)

jreback commented:
can you show what this does for a truncated repr? e.g. > 10 groups.

JulianWgs commented:
> can you show what this does for a truncated repr? e.g. > 10 groups.

I've updated the original pull request text accordingly

jreback commented on pandas/core/groupby/groupby.py:
```diff
@@ -548,6 +549,29 @@ def __repr__(self) -> str:
         # TODO: Better repr for GroupBy object
         return object.__repr__(self)
 
+    def _repr_html_(self) -> str:
```
I'd like to reorg this to use a GroupbyFormatter located in pandas/io/formats/groupby.py (it can do pretty much this but just locate the code there) as this is where we keep all of the formatting code.

could *also* add a `.to_string()` method but not sure that's actually worth it (maybe open an issue for that).

JulianWgs replied:
Thank for the review! Do you mean pandas/io/formats/html.py? Should I add a new function and then just call that function from the above location?

jreback replied:
no, i mean pandas/io/formats/format.py (ok to just shove in there is fine, we should split that file up but that's for later).

JulianWgs committed:
Move html repr function to io/formats/format.py

- from groupby module
- https://github.com/pandas-dev/pandas/pull/34926#discussion_r483957300

JulianWgs committed:
Add doc string and return type annotation

JulianWgs commented:
Is there a DataFrameGroupBy type?

jreback commented:
> Is there a DataFrameGroupBy type?

yes that's the class where you did the change (or you may have done it in the super class)

JulianWgs committed:
Add type annotations for input arg

JulianWgs committed:
Fix linting errors

JulianWgs committed:
Move import to the correct location

JulianWgs committed:
Remove pandas type annotations

because of circular import
https://github.com/pandas-dev/pandas/pull/34926/checks

JulianWgs committed:
Merge remote-tracking branch 'upstream/master'

JulianWgs committed:
Remove inconsistent use of pd namespace in tests

- https://github.com/pandas-dev/pandas/pull/34926/checks?check_run_id=1309077928

rhshadrach commented on pandas/io/formats/format.py:
```diff
@@ -2020,3 +2020,39 @@ def buffer_put_lines(buf: IO[str], lines: List[str]) -> None:
+
+def repr_html_groupby(group_obj) -> str:
+    """
+    Create a HTML representation for a grouped dataframe or series.
```
nit: an HTML representation, since "aitch" starts with a vowel.

Also, DataFrame or Series capitalization (and throughout the docstring)

JulianWgs replied:
I did not know that :)

rhshadrach commented on pandas/io/formats/format.py:
```diff
@@ -2020,3 +2020,39 @@ def buffer_put_lines(buf: IO[str], lines: List[str]) -> None:
+    Parameters
+    ----------
+    group_obj : [DataFrameGroupBy, SeriesGroupBy]
+        A grouped dataframe or a series of a grouped dataframe.
```
You're describing the type twice here; should be something like "Object to make HTML representation of."

rhshadrach commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2059,6 +2060,93 @@ def test_groups_repr_truncates(max_seq_items, expected):
+def test_groupby_repr():
+    """
+    All groups and all rows in a group are shown in html output.
+    """
```
No docstring on tests; instead add comment with issue number (or PR if no issue exists):

    # GH 34926 - All groups and all rows in a group are shown in html output.

Same with other tests.

rhshadrach commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2059,6 +2060,93 @@ def test_groups_repr_truncates(max_seq_items, expected):
+
+    dfs_from_html = pd.read_html(StringIO(html_groupby), index_col=0)
+
+    # For each group only test first and last row
```
Comment on _why_ you're only testing first/last rows.

rhshadrach commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2059,6 +2060,93 @@ def test_groups_repr_truncates(max_seq_items, expected):
+
+    dfs_from_html = pd.read_html(StringIO(html_groupby), index_col=0)
+
+    # Test first and last group
```
Same as above.

JulianWgs replied:
The comment is almost exactly the same as in test_groupby_repr_truncated_group. Should I rather refer to that comment?

rhshadrach reviewed (changes requested):
Changes look good, nice job on the tests. A few style requests.

rhshadrach commented:
@jreback: Any thoughts on whether tests for `groupby._repr_html_` should be in groupby tests or `io.formats.test_to_html` (or somewhere else)?

JulianWgs committed:
Fix typo and capitalize pandas objs correctly

JulianWgs committed:
Change docstring to comment in groupby repr test

- add link to Github Pull Request

JulianWgs committed:
Add additional explanation in groupby_repr test

- why only the first/last rows/groups are tested

rhshadrach reviewed (changes requested):
I don't think checking only the first/last row/group is sufficient; we should be checking that exactly the right ones are included.

It seems feasible to combine these into one test by parametrizing the test with n_groups, n_rows, check_n_rows, check_n_groups (so the number of rows/groups to check is hard coded as a parameter), or names of that sort. If necessary, can use something like "all" if separate logic is needed when no rows/groups are hidden.

jreback commented on pandas/io/formats/format.py:
```diff
@@ -2020,3 +2020,39 @@ def buffer_put_lines(buf: IO[str], lines: List[str]) -> None:
     buf.write("\n".join(lines))
+
+
+def repr_html_groupby(group_obj) -> str:
```
this should use the same machines as DataFrameFormatter/DataFrameRenderer (subclass as appropriate), which was recently changed).

JulianWgs replied:
Sorry, for the long inactivity. I don't get how I would use the DataFrameFormatter? Is there documentation on this?

JulianWgs replied:
Sorry, for coming back to this again, but I really dont get what code I should change or how? Could you tell me which line in my code I have to rewrite?

JulianWgs replied:
Is this still relevant? I still would need some guidance :)

JulianWgs committed:
Test more rows in groupby repr when truncated

JulianWgs committed:
Test more groups in groupby repr when truncated

arw2019 reviewed (commented):
@JulianWgs if you can address comments. There's also a minor pre-commit failure

JulianWgs committed:
Refactor groups repr html

- combine all previous test cases into one parametrized test case

arw2019 commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2059,6 +2060,49 @@ def test_groups_repr_truncates(max_seq_items, expected):
+        }
+    )
+
+    df_groupby = df.groupby("C")
```
Personal opinion (feel free to ignore) but I find these names confusing. I'd use `gb` for `df_groupby`, `gb_html` for `html_groupby`, `gb_html_as_frame` for `df_from_html`, `gb_iter` for `df_group_iter`

arw2019 reviewed (changes requested):
Thanks for picking this up!

In addition to addressing comments you'll want to merge master to make sure that this is up to date, and add a whatsnew (1.3, under enhancements)

JulianWgs committed:
Merge remote-tracking branch 'upstream/master'

JulianWgs committed:
Add whatsnew entry for group-by HTML representation

JulianWgs committed:
Fix test case name

JulianWgs committed:
Rename groupby objects

JulianWgs committed:
Add case for single and tuple groupby key

arw2019 reviewed (commented):
One more comment - we might want an example for this in the groupby section of the user guide (but could be left for a followon too)

JulianWgs committed:
Merge remote-tracking branch 'upstream/master'

simonjayhawkins commented:
@JulianWgs can you resolve conflicts (and move release note to 1.4)

JulianWgs committed:
Merge branch 'master' into master

JulianWgs committed:
Move whats new to 1.4.0 release

JulianWgs commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2059,6 +2060,97 @@ def test_groups_repr_truncates(max_seq_items, expected):
+    for k, (group_name, df_group) in enumerate(df_groupby):
+        dtype = df_group.iloc[0].dtype
+        tm.assert_series_equal(
+            dfs_from_html[k].iloc[0].astype(dtype), df_group.iloc[0], check_names=False
```
New code should be much clearer :) HTML is parsed as string type and needs to converted to integer manually

JulianWgs commented:
@simonjayhawkins Done :)

jreback commented:
this is quite old, happen to reopen if actively worked on.
